File: codeforces/2020_11_28_practice/B.py

import copy
from collections import defaultdict
from collections import deque
import math
import sys
import pickle
import bz2
import base64

if sys.version_info[0] == 2:
    input = raw_input

def nn():
    return int(input())

def li():
    return list(input())

def mi():
    return map(int, input().split())

def lm():
    return list(map(int, input().split()))

def compress(abc):
    return base64.b64encode(
        bz2.compress(
            pickle.dumps(
                abc
            )
        )
    )

def expand(abc):
    return pickle.loads(
        bz2.decompress(
            base64.b64encode(
                abc
            )
        )
    )


def solve(rows, cols, M, N):
    '''
    Approach:
    - Pick any row, find its first element in a column somewhere. That's now col 1.
    - slot in all the other rows based on where their first element is in column 1.
    '''

    # Find where the 1st element of rows[0] appears in cols.
    fcol = None
    for first_col_idx in range(len(cols)):
        for q in cols[first_col_idx]:
            if q == rows[0][0]:
                fcol = cols[first_col_idx]
                break
        if fcol is not None:
            break

    # Need inverse map for the special column.
    fcol_inv = {
        val: idx
        for idx, val in enumerate(fcol)
    }

    out = [[] for k in range(len(rows))]
    for q in rows:
        # where does rows[k] go?
        pos = fcol_inv[q[0]]
        out[pos] = q

    return out

T = nn()
for testID in range(1, T+1):
    M, N = lm()
    rows = []
    for k in range(M):
        rows.append(lm())
        assert len(rows[-1]) == N
    cols = []
    for k in range(N):
        cols.append(lm())
        assert len(cols[-1]) == M

    out = solve(rows, cols, M, N)
    print(
        '\n'.join(
            ' '.join(
                str(q) for q in row
            )
            for row in out
        )
    )

File: codeforces/2020_11_28_practice/B.cpp

#include <iostream>
#include <vector>
#include <map>

void solve(
    const std::vector< std::vector<int> >& rows,
    const std::vector< std::vector<int> >& cols,
    std::vector< std::vector<int> >& out
)
{
    int R = rows.size();
    int C = cols.size();

    // Find where the 1st element of rows[0] appears in cols.
    int special_col = -1;
    for ( int r = 0 ; r < rows.size() ; ++r ) {
        for ( int c = 0 ; c < cols.size() ; ++c ) {
            if ( cols[ c ][ r ] == rows[ 0 ][ 0 ] ) {
                special_col = c;
                break;
            }
        }
        if ( special_col >= 0 ) {
            break;
        }
    }

    // need inverse map for the special column.
    std::map<int, int> scol_inv;
    for ( int r = 0 ; r < R ; ++r ) {
        scol_inv[ cols[ special_col ][ r ] ] = r;
    }

    for ( int r = 0 ; r < R ; ++r ) {
        int idx = scol_inv[ rows[ r ][ 0 ] ];
        for ( int c = 0 ; c < C ; ++c ) {
            out[ idx ][ c ] = rows[ r ][ c ];
        }
    }
    return;
}

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie( NULL );
    std::cout.tie( NULL );

    int T;
    std::cin >> T;
    std::vector< std::vector<int> > rows;
    std::vector< std::vector<int> > out;
    std::vector< std::vector<int> > cols;
    int R, C;
    for ( int k = 0 ; k < T ; ++k ) {
        std::cin >> R >> C;
        int M = R;
        int N = C;
        rows.resize( M );
        out.resize( M );
        for ( int r = 0 ; r < M ; ++r ) {
            rows[ r ].resize( N );
            out[ r ].resize( N );
            for ( int c = 0 ; c < N ; ++c ) {
                std::cin >> rows[ r ][ c ];
            }
        }

        cols.resize( N );
        for ( int c = 0 ; c < N ; ++c ) {
            cols[ c ].resize( M );
            for ( int r = 0 ; r < M ; ++r ) {
                std::cin >> cols[ c ][ r ];
            }
        }
        solve(rows, cols, out);

        for ( int r = 0 ; r < R ; ++r ) {
            for ( int c = 0 ; c < C ; ++c ) {
                std::cout << out[r][c] << ' ';
            }
            std::cout << '\n';
        }
    }

    return 0;
}
